### PR TITLE
Ore dict support

### PR DESCRIPTION
**Pull request by Techlone**
So, I created the basic action class for AddRecipeAction per machine (now only AlloySmelter in a business).
This class iterates for IOreDictEntry inputs and creates and fills a 'matrix' of recipe's parameters (inputs, outputs, ticks, EU, special values - always).
For example: AlloySmelter has 2 inputs and one output. All outputs always must be an IItemStack (or Fluid**\* don't remember).

```
var oreCopper  = <ore:ingotCopper>;
var oreTin = <ore:ingotTin>;
var gtBronze = <ore:ingotBronze>.greggetItem;
AlloySmelter.addRecipe(gtBronze, oreCopper * 3, oreTin);
```

(`greggetItem` #6 not merged into this branch yet)
It should make a la cross join between inputs.

Also I maked some refactoring. Moved classes in machines subpackage. I think it's a better place for them.

I need a code review. Ping @jaredlll08 @leagris
Can you spend a few minutes for me please? :) Maybe it looks too difficult or like a shitcode, huh?

### CODE REVIEW

**Comment by Techlone** [comment]
@Dream-Master 
Do not try to merge now, ok?

File: src/main/java/gttweaker/mods/gregtech/machines/AlloySmelter.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+        @Override
+        protected void applySingleRecipe(Object... args) {
+            RA.addAlloySmelterRecipe(
+                    (ItemStack) args[0],
+                    (ItemStack) args[1],
+                    (ItemStack) args[2],
+                    (Integer) args[3],
+                    (Integer) args[4]);
+        }
```
Comment (Techlone) [inline src/main/java/gttweaker/mods/gregtech/machines/AlloySmelter.java]: The order in the call of the superclass constructor is important.
I wanted to include the reflection, but I think this is a bad idea.

**Comment by leagris** [comment]
@Techlone you must iterate IIngredient.getitems list, this will include any ingredient items contained in the list, including those referring ore dictionary entries.

**Comment by Dream-Master** [comment]
@Techlone 
Sorry some times I sleep ;-). Guess I living in a different time zone then you ;-). Will look into it when I came from work.

**Comment by Techlone** [comment]
@leagris
Hmm, maybe... But I saw implementations of IIngredient and only IOreDictEntry returns not itself.
I missed something?
[As you can see](https://github.com/Techlone/Minetweaker-Gregtech-5-Addon/commit/1697a1d6f22a71c22d05ded91b69d10262162e4c#diff-ecb95a8fea6c39372a56fa2db8d4e4f9R29) I iterate them.

UPD: @leagris yep, you right, I don't thought about the amount (`IngredientStack` class) and other IIngredient implementations. Will be think how to resolve it.

**Comment by Techlone** [comment]
Maybe I explain bad. See my idea

The example from the head.

```
Ore entries for <ore:ingotTin> :
    <IC2:itemIngot:1>
    <Forestry:ingotTin>
    <Railcraft:ingot:2>

Ore entries for <ore:ingotCopper> :
    <IC2:itemIngot>
    <Forestry:ingotCopper>
    <Railcraft:ingot:1>
```

Then I'm trying make that matrix (List of Lists)

```
    <IC2:itemIngot:1>   + <IC2:itemIngot>
    <Forestry:ingotTin> + <IC2:itemIngot>
    <Railcraft:ingot:2> + <IC2:itemIngot>
    <IC2:itemIngot:1>   + <Forestry:ingotCopper>
    <Forestry:ingotTin> + <Forestry:ingotCopper>
    <Railcraft:ingot:2> + <Forestry:ingotCopper>
    <IC2:itemIngot:1>   + <Railcraft:ingot:1>
    <Forestry:ingotTin> + <Railcraft:ingot:1>
    <Railcraft:ingot:2> + <Railcraft:ingot:1>
```

After that I add recipe per each combination. Who think "it's wrong"?

Of course I can add tons cycles in the overrided `apply` method per each machine. BRUTALITY!

**Comment by Dream-Master** [comment]
I cant merge it without conflicts. Why you not make a own branch for it because it is wip i guess?

**Comment by Techlone** [comment]
> Why you not make a own branch

I did that in my fork (feature/oredict-support). In local I resolved conflicts and after testing I will push the result.
I created this pull request for the code review. So, now I'm testing. And I said

> Do not try to merge now, ok?

I will say you when I finish.

**Comment by Dream-Master** [comment]
sorry i am a java noob so be patient with me :)

**Comment by Techlone** [comment]
The progress. The script

```
ArcFurnace.addRecipe([oreDustCopper.firstItem * 4, oreDustGold.firstItem], <ore:ingotRoseGold> * 5, <liquid:oxygen> * 144, [10000, 10000], 20, 300);
```

gave me a correct result. The `fistItem` it's my [feature](https://github.com/jaredlll08/CraftTweaker/pull/8) which not merged yet.
It needs to be tested.

**Comment by Dream-Master** [comment]
Why the first item function  is better than only normal ore support or what it do different?

**Comment by Techlone** [comment]
@Dream-Master
It's different. Now I think it was unnecessary field. For GT's things you should use `greggedItem`. It ensures that the item will be from GT mod. I hope :)
Also the `firstItem` may be usefull for other mods or something else. Imagination...

I need to change all machine's classes and test them. After that you can make a beta release
image

**Comment by Dream-Master** [comment]
@Techlone 
shall i merge this or wait you are done ?

**Comment by Techlone** [comment]
Please wait.

**Comment by Dream-Master** [comment]
ok. np.

**Comment by Techlone** [comment]
Sorry for waiting I was busy.
Push will be today/tomorrow.

**Comment by Dream-Master** [comment]
:+1:

**Comment by Techlone** [comment]
So...
You can merge it. Please make a **beta** release. It must be tested.

**Comment by Dream-Master** [comment]
I make a beta release.

@Techlone 
What i can test here only if ordict support works in all the machines like ore:ingotIron or any thing else?

**Comment by Techlone** [comment]
I not sure that the ore dict support works perfectly.
Test should look like so

```
AlloySmelter.addRecipe(<ore:ingotRoseGold>.greggedItem * 5, <ore:dustCopper> * 4, <ore:dustGold>, 100, 220);
```

**Comment by Dream-Master** [comment]
@Techlone 
You sure outputs can be ordicted. How the machine knows what have to output?

**Comment by Techlone** [comment]
Ehm, yes my bad. See the exmaple again. Output can't be an oredict.

**Comment by Dream-Master** [comment]
Need this to be fixed? Yesterday i dont have the time to test it.

**Comment by Techlone** [comment]
I made mistake only in the testing script example above (see `greggedItem`). Are you want to fix something more? :)

**Comment by Dream-Master** [comment]
greggedItem meens ore or dust or item from Greg i guess?

**Comment by Techlone** [comment]
It means a first GT's item from an oredict. I think it's usefull.
`<gregtech:gt.metaitem.BLAH-BLAH-BLAH>` very hard to read and write.
Also if the oredict has multiple GT's items it still works but it write an error message. #6

**Comment by Dream-Master** [comment]
Ah ok. Will try your version later then.

**Pull request merged**